### ISSUE TITLE
Add negative index regression test for json-patch

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#120563]
https://github.com/kubernetes/kubernetes/pull/120327 reverted json-patch back to 4.12.0 to fix a regression in negative index json-patch behavior

A test that exercises the bug fixed in https://github.com/kubernetes/kubernetes/pull/105896 is needed to ensure we don't accidentally bump again and lose the fix

_Originally posted by @liggitt in https://github.com/kubernetes/kubernetes/issues/120326#issuecomment-1701891191_
           
cc @Jefftree @soltysh

**Comment by MeenuyD** [comment]
Hello @liggitt I would like to work on this issue can i work on this issue

**Comment by liggitt** [comment]
sounds good, a unit test that tries to apply a patch like the one indicated in https://github.com/kubernetes/kubernetes/issues/105895#issue-1035786694 would probably be sufficient

**Comment by MeenuyD** [comment]
I am confused about whether should I change the version of json-patch.

**Comment by liggitt** [comment]
No, we already reverted json-patch to fix the regression. This issue is to add a new unit test exercising a negative index `replace` json-patch as described in https://github.com/kubernetes/kubernetes/issues/105895#issue-1035786694 to make sure we don't reintroduce the regression

**Comment by MeenuyD** [comment]
Ok thanks for guiding

**Comment by carlory** [comment]
/assign MeenuyD

I assign this issue to you in order to avoid someone to do duplicated work.

**Comment by MeenuyD** [comment]
Hello @liggitt should we have to add a replacement operation for the negative index because issue #105895 says this is resolved by changing the version from 4.11.0 to 4.12.0 Can you please explain more

**Comment by liggitt** [comment]
that bumped the dependency, fixing the issue, but we never added a test to exercise the fix. Then, when json-patch accidentally got bumped to a version that regressed the fix, no test caught it

**Comment by aryabhatta-dey** [comment]
Hey! There has been no activity on this for some time. I will like to take a stab at this if nobody's working on it @liggitt @carlory

**Comment by aryabhatta-dey** [comment]
Hey! I will like to take a stab at this since it looks like nobody is actively working on it.
/assign

**Comment by Vandit1604** [comment]
I don't see anyone actively working on this issue. I would like to give it a try.
/assign 

@liggitt I searched the test file for the `JsonPatch` and stumbled on this testing function in [helpers_test.go](https://github.com/kubernetes/kubernetes/blob/6aac45ff1e99068e834ba3b93b673530cf62c007/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go#L249-L309). 
I want to ask if adding a testcase with negative index in the above function would count as a solution. If yes, I can submit a Pull Request.

Something like this 
```go
{
	obj:       &corev1.Pod{},
	fragment:  `[ {"op": "replace", "path": "/spec/volumes/-1/foo", "value": "bar"} ]`,
	expectErr: true,
},
```

**Comment by liggitt** [comment]
@Vandit1604 I'm not 100% sure without digging, but if adding that negative index test case passes on master and fails on v1.28.0, that seems like what we're looking for

**Comment by Vandit1604** [comment]
I saw interesting behaviour in branch `release-1.28` [helpers_test.go's TestJSONPatch function](https://github.com/kubernetes/kubernetes/blob/release-1.28/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go#L249). When I added this testcase
```go
{
			obj: &corev1.Pod{
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "test"},
				},
			},
			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
			expected: &corev1.Pod{
				TypeMeta: metav1.TypeMeta{
					Kind:       "Pod",
					APIVersion: "v1",
				},
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "foo"},
				},
				Spec: corev1.PodSpec{},
			},
			expectErr: true,
}
```
I expected it to fail on negative index replace operation but it didn't and gave 
```bash
root@vandit-box:/home/vandit/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/util# go test
--- FAIL: TestJSONPatch (0.00s)
    helpers_test.go:329: testcase[3], unexpected non-error
FAIL
exit status 1
FAIL	k8s.io/kubectl/pkg/cmd/util	0.014s
```

I also ran a `make all` it also didn't fail on `release-1.28`
```bash
--- FAIL: TestJSONPatch (0.00s)
    helpers_test.go:329: testcase[3], unexpected non-error
FAIL
FAIL	k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util
```

I ran tests on master after adding 
```go
{
			obj: &corev1.Pod{
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "test"},
				},
			},
			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
			expected: &corev1.Pod{
				TypeMeta: metav1.TypeMeta{
					Kind:       "Pod",
					APIVersion: "v1",
				},
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "foo"},
				},
				Spec: corev1.PodSpec{},
			},
			// expectErr: true,
}
```
And every test passed. 

Was this the desired output ? Since I am assuming the unexpected error to be false alarm or It should be like this?

**Comment by liggitt** [comment]
The negative index should *not* error, right?

**Comment by Vandit1604** [comment]
Thanks for pointing that out. I ran the tests again both on `master` and `release-1.28`. All the testcases passed on both branches.

I saw that we are only using the JsonPatch() function only in 2 more files i.e 
```go
staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go            func JSONPatch(codec runtime.Codec, dst runtime.Object, fragment string) (runtime.Object, error) {
staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go       out, err := JSONPatch(codec, test.obj, test.fragment)
staging/src/k8s.io/kubectl/pkg/cmd/util/override_options.go   return JSONPatch(codec, obj, o.Options.Overrides)
```
I think I'll need to look at other places as it seems we're not using the JSONPatch() function for patching the resources anywhere.

**Comment by Vandit1604** [comment]
TLDR: The `kubectl patch` replace operation was working fine with negative index on **release-1.28**

I created a nginx-deployment and added 3 more `ports` in `spec.container.ports` with

```bash
kubectl patch deployment nginx-deployment --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/ports/-", "value": {"containerPort": 10253,"name": "metrics2"}}]'
```

<details open>
<summary>Ran the above command 2 more times with different values.</summary>
 <br>
 
 ```yaml
            "spec": {
                "containers": [
                    {
                        "image": "nginx:1.14.2",
                        "imagePullPolicy": "IfNotPresent",
                        "name": "nginx",
                        "ports": [
                            {
                                "containerPort": 80,
                                "protocol": "TCP"
                            },
                            {
                                "containerPort": 10254, # we will change this value i.e -3 index
                                "name": "metrics",
                                "protocol": "TCP"
                            },
                            {
                                "containerPort": 10254,
                                "name": "metrics1",
                                "protocol": "TCP"
                            },
                            {
                                "containerPort": 10253,
                                "name": "metrics2",
                                "protocol": "TCP"
                            }
                        ],
                        "resources": {},
                        "terminationMessagePath": "/dev/termination-log",
                        "terminationMessagePolicy": "File"
                    }
                ],
                "dnsPolicy": "ClusterFirst",
                "restartPolicy": "Always",
                "schedulerName": "default-scheduler",
                "securityContext": {},
                "terminationGracePeriodSeconds": 30
            }
        }
    },
```

 </details>
 
<details open>
<summary> Then the kubectl replace patch worked perfectly fine. Here's the output</summary>
<br>

 ```yaml
 root@vandit-box:/home/vandit/go/src/k8s.io/kubernetes# k get deployment nginx-deployment -o json
{
 ...
   "spec": {
                "containers": [
                    {
                        "image": "nginx:1.14.2",
                        "imagePullPolicy": "IfNotPresent",
                        "name": "nginx",
                        "ports": [
                            {
                                "containerPort": 80,
                                "protocol": "TCP"
                            },
                            {
                                "containerPort": 10251,
                                "name": "metrics",
                                "protocol": "TCP"
                            },
                            {
                                "containerPort": 10254,
                                "name": "metrics1",
                                "protocol": "TCP"
                            },
                            {
                                "containerPort": 10253,
                                "name": "metrics2",
                                "protocol": "TCP"
                            }
                        ],
                        "resources": {},
                        "terminationMessagePath": "/dev/termination-log",
                        "terminationMessagePolicy": "File"
                    }
                ],
                "dnsPolicy": "ClusterFirst",
                "restartPolicy": "Always",
                "schedulerName": "default-scheduler",
                "securityContext": {},
                "terminationGracePeriodSeconds": 30
            }
        }
    },
...
}
 ```


</details>
 
 ```bash
 root@vandit-box:/home/vandit/go/src/k8s.io/kubernetes# kubectl patch deployment nginx-deployment --type='json' -p='[{"op": "replace", "path": "/spec/template/spec/containers/0/ports/-3", "value": {"containerPort": 10251,"name": "metrics"}}]'
deployment.apps/nginx-deployment patched
```

Kubectl Version. I was testing using KinD
```bash
root@vandit-box:/home/vandit/go/src/k8s.io/kubernetes# kubectl version
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.5-5+5d19345cef50b0-dirty
```

**Comment by Vandit1604** [comment]
If we want to make kubernetes durable against index out of range for Kubectl Patch, Adding the following test case in [helpers_test.go's TestJSONPatch function](https://github.com/kubernetes/kubernetes/blob/release-1.28/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers_test.go#L249) could be a possible solution.

```go
{
			obj: &corev1.Pod{
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "test"},
				},
			},
			fragment: `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`,
			expected: &corev1.Pod{
				TypeMeta: metav1.TypeMeta{
					Kind:       "Pod",
					APIVersion: "v1",
				},
				ObjectMeta: metav1.ObjectMeta{
					Name:       "foo",
					Finalizers: []string{"foo", "bar", "foo"},
				},
				Spec: corev1.PodSpec{},
			},
}
```

**Comment by liggitt** [comment]
oh, I missed that unit test was only for kubectl's json-patch use... the apiserver also does json patches, there might be negative index tests that could be added there

**Comment by Vandit1604** [comment]
These are the [results that came up for replace operation](https://github.com/search?q=repo%3Akubernetes%2Fkubernetes+%22op%22%3A+%22replace%22&type=code) 
and these are the [results under apiserver](https://github.com/search?q=repo%3Akubernetes%2Fkubernetes+%22op%22%3A+%22replace%22+path%3A%2Fapiserver%2F&type=code) 

These are the places where we are using replace operation on positional arrays in the codebase. Should the testcase be added everywhere? I can add commits in this Pull request I have created https://github.com/kubernetes/kubernetes/pull/122625 

https://github.com/kubernetes/kubernetes/blob/60dcf7fd8dbbb05fa02a08217001f98c5b1507c2/staging/src/k8s.io/kubectl/pkg/cmd/patch/patch.go#L104

https://github.com/kubernetes/kubernetes/blob/60dcf7fd8dbbb05fa02a08217001f98c5b1507c2/staging/src/k8s.io/kubectl/pkg/cmd/run/run_test.go#L725

https://github.com/kubernetes/kubernetes/blob/60dcf7fd8dbbb05fa02a08217001f98c5b1507c2/test/e2e/apimachinery/webhook.go#L471

**Comment by liggitt** [comment]
> These are the places where we are using replace operation on positional arrays in the codebase.

JSON patches are accepted by kube-apiserver, so end-users can specify replace operations. That's the main thing we want to make sure works properly.

We should expand the staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go#TestJSONPatch unit test to exercise successful negative index patching.

**Comment by Vandit1604** [comment]
I added this testcase in [rest_test.go](https://github.com/kubernetes/kubernetes/blob/9e0eccabb47d9044b8ec72ff27df848692b88823/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go#L224-L246)

```go
{
	name:  "valid-negative-index-patch",
	patch: `[{"op": "replace", "path": "/spec/containers/-1/image", "value": "foo"}]`,
},
```

When I run `go test`, It fails with
```bash
--- FAIL: TestJSONPatch (0.00s)
    rest_test.go:269: valid-negative-index-patch: expect no error when applying json patch, but got the server rejected our request due to an error in our request
FAIL
exit status 1
FAIL	k8s.io/apiserver/pkg/endpoints/handlers	0.025s
```

Although When I try to patch the Pods with
```bash
kubectl patch pod nginx-deployment-86dcfdf4c6-f8p2s --type='json' -p='[{"op": "replace", "path": "/spec/containers/-1/image", "value": "hello-world"}]'
```
where the path is exactly same, it works fine. I also tried 
```
 `[ {"op": "replace", "path": "/metadata/finalizers/-1", "value": "foo"} ]`
 ```
 But the same `the server rejected our request due to an error in our request`

On further digging, It seems to me the pod doesn't contain a `container` object. (Assuming)
when I printed the pod

```&Pod{ObjectMeta:{podA      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:PodSpec{RestartPolicy:,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,NodeSelector:map[string]string{},ServiceAccountName:,DeprecatedServiceAccount:,NodeName:,HostNetwork:false,HostPID:false,HostIPC:false,Hostname:,Subdomain:,SchedulerName:,},Status:PodStatus{Phase:,Conditions:[]PodCondition{},Message:,Reason:,HostIP:,PodIP:,StartTime:<nil>,},}```

and there's no slice present in [v1/types.go](https://github.com/kubernetes/kubernetes/blob/9e0eccabb47d9044b8ec72ff27df848692b88823/staging/src/k8s.io/apiserver/pkg/apis/example/v1/types.go) and the slices in [meta/v1/types.go](https://github.com/kubernetes/kubernetes/blob/9e0eccabb47d9044b8ec72ff27df848692b88823/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go) have `patchStrategy:"merge"`

Is there something, I'm missing ?

**Comment by Vandit1604** [comment]
I was not giving a value to any of the array that's why it was always failing. I have implemented the required changes and committed them in #122625. I think the PR is ready for review. @liggitt

**Comment by Vandit1604** [comment]
Fixed by #122625
This issue can be closed.
